cmd/notation/manifest.go

```diff
@@ -38,59 +38,66 @@ func resolveReferenceWithWarning(ctx context.Context, inputType inputType, refer
 // resolveReference resolves user input reference based on user input type.
 // Returns the resolved manifest descriptor and resolvedRef in digest
 func resolveReference(ctx context.Context, inputType inputType, reference string, sigRepo notationregistry.Repository, fn func(string, ocispec.Descriptor)) (ocispec.Descriptor, string, error) {
+	repositoryRef, tagOrDigestRef, err := parseReference(reference, inputType)
+	if err != nil {
+		return ocispec.Descriptor{}, "", err
+	}
+	manifestDesc, err := getManifestDescriptor(ctx, tagOrDigestRef, sigRepo)
+	if err != nil {
+		return ocispec.Descriptor{}, "", fmt.Errorf("failed to get manifest descriptor: %w", err)
+	}
+	resolvedRef := repositoryRef + "@" + manifestDesc.Digest.String()
+	if _, err := digest.Parse(tagOrDigestRef); err == nil {
+		// tagOrDigestRef is a digest reference
+		if tagOrDigestRef != manifestDesc.Digest.String() {
+			// tagOrDigestRef does not match the resolved digest
+			return ocispec.Descriptor{}, "", fmt.Errorf("user input digest %s does not match the resolved digest %s", tagOrDigestRef, manifestDesc.Digest.String())
+		}
+		return manifestDesc, resolvedRef, nil
+	}
+	// tagOrDigestRef is a tag reference
+	if fn != nil {
+		fn(tagOrDigestRef, manifestDesc)
+	}
+	return manifestDesc, resolvedRef, nil
+}
+
+// parseReference parses the user input reference based on the input type.
+// Returns the repository reference and tag or digest reference.
+func parseReference(reference string, inputType inputType) (repositoryRef, tagOrDigestRef string, err error) {
```

Comment:
nit: code style: try to avoid named return values. document them in the documentation instead.
```suggestion
func parseReference(reference string, inputType inputType) (string, string, error) {
```

Reply:
Updated.